File: Chapter 03/Python/section_3_3_1_srp/before/player.py
"""
ANTI-PATTERN: The "God Class" (SRP Violation).

ARCHITECTURE PROBLEM: This class is doing way too much. According to the 
Single Responsibility Principle, a class should have only ONE reason to change. 

This Python implementation currently has three distinct reasons:
1. Domain Logic: If the physical rules of how a player dribbles change.
2. Tactical Logic: If the AI algorithm for positioning is updated.
3. Persistence Logic: If the database schema or the way we save data changes.

In a professional Python environment, this makes unit testing a nightmare 
because you cannot test a simple 'dribble' without the class carrying around 
unused (and potentially broken) database code.
"""

class Player:
    def __init__(self, name: str):
        self.name = name

    # Responsibility 1: Player’s own state/abilities (Domain Logic)
    # This is the only logic that truly belongs inside this entity.
    def dribble_ball(self):
        print(f"  [Action] {self.name} is dribbling the ball down the court.")

    # 🚨 ARCHITECTURE WARNING: Tactical Logic (Should be in a Service)
    # This couples the Player to a specific AI strategy.
    def determine_best_position(self):
        print(f"  [Tactics] Calculating optimal court position for {self.name}...")

    # 🚨 ARCHITECTURE WARNING: Data Persistence (Should be in a Repository)
    # This couples the domain model to a specific infrastructure or database.
    def save_stats_to_database(self):
        print(f"  [Database] Saving {self.name}'s game stats to the database.")

File: Chapter 03/JS/section_3_3_1_srp/before/player.js
/**
 * ANTI-PATTERN: The "God Class" (SRP Violation).
 * * ARCHITECTURE PROBLEM: This class is doing too much. According to the Single 
 * Responsibility Principle, a class should have only one reason to change. 
 * This class currently has THREE:
 * * 1. If the physical rules of the game change (Action).
 * 2. If the AI algorithms change (Tactics).
 * 3. If the database schema or persistence logic changes (Persistence).
 * * In a JavaScript environment, this coupling makes unit testing difficult because 
 * you might just want to test a dribble move, but you're forced to carry around 
 * code that expects a database connection.
 */
class Player {
    constructor(name) {
        this.name = name;
    }

    /**
     * Responsibility 1: Domain Logic. 
     * (This is the core purpose of the entity.)
     */
    dribbleBall() {
        console.log(`  [Action] ${this.name} is dribbling the ball down the court.`);
    }

    /**
     * Responsibility 2: Tactical Logic.
     * (Should be moved to a dedicated service or strategy.)
     */
    determineBestPosition() {
        console.log(`  [Tactics] Calculating optimal court position for ${this.name}...`);
    }

    /**
     * Responsibility 3: Data Persistence.
     * (Strongly couples the domain model to infrastructure.)
     */
    saveStatsToDatabase() {
        console.log(`  [Database] Saving ${this.name}'s game stats to the database.`);
    }
}

module.exports = Player;